### PR TITLE
Clarify that HTTP self-descriptive messages have an exception …

### LINKED ISSUE #700: Messages aren't always self-descriptive
Section 6 claims that [messages are self-descriptive](https://httpwg.org/http-core/draft-ietf-httpbis-semantics-latest.html#rfc.section.6.p.5).  It uses the word "intended", which is an intention that is broadly achieved, with only one exception that I'm aware of at the syntactic level.

Having recently implemented message/http, I discovered that some response messages cannot be processed without knowledge of the corresponding request.  This is pretty annoying and might be worth noting as a possible exception here.  As noted in [Section 6.3 of -messaging](https://httpwg.org/http-core/draft-ietf-httpbis-messaging-latest.html#message.body.length) the response to a HEAD request might include Content-Length, but the content of the response is always empty.

At a semantic level, I don't think that any response can be considered without knowledge of the request, so I don't think that self-descriptiveness can possibly apply there.  Besides, this section is clearly intended to provide an abstract definition of the syntax and not semantics.

### PR DESCRIPTION
**Pull request by royfielding**
…  in that the request must be understood in order to parse and interpret the response.

Fixes #700

### CODE REVIEW

File: draft-ietf-httpbis-semantics-latest.xml [hunk 1]
```diff
@@ -2169,7 +2169,13 @@ Sun Nov  6 08:49:37 1994         ; ANSI C's asctime() format
...
    have been compressed or elided in transit, without requiring an
    understanding of the sender's current application state (established via
-   prior messages).
+   prior messages). However, there is one significant exception: each
+   response relies on prior understanding of its corresponding request.
+   A client might need to remember the method, target URI, HTTP version, and
+   content negotiation fields that it used for a given request in order to
+   correctly parse, interpret, or cache the response. For example, responses
+   to the <x:ref>HEAD</x:ref> method look just like the beginning of a
+   response to <x:ref>GET</x:ref>, but cannot be parsed the same.
```
Comment (mnot) [inline draft-ietf-httpbis-semantics-latest.xml:2178]:
```suggestion
   prior messages). However, a response must be understood within the context
   of its corresponding request in order to
   correctly parse, interpret, or cache the response. For example, responses
   to the <x:ref>HEAD</x:ref> method look just like the beginning of a
   response to <x:ref>GET</x:ref>, but cannot be parsed in the same manner.
```

**Review by mnot — APPROVED** [review]
This is fine, but can be simplified.

**Review by reschke — APPROVED** [review]

File: draft-ietf-httpbis-semantics-latest.xml [hunk 2]
```diff
@@ -2169,7 +2169,11 @@ Sun Nov  6 08:49:37 1994         ; ANSI C's asctime() format
    looking at the message itself, after decoding or reconstituting parts that
    have been compressed or elided in transit, without requiring an
    understanding of the sender's current application state (established via
-   prior messages).
+   prior messages). However, a response must be understood within the context
+   of its corresponding request in order to
+   correctly parse, interpret, or cache the response. For example, responses
+   to the <x:ref>HEAD</x:ref> method look just like the beginning of a
+   response to <x:ref>GET</x:ref>, but cannot be parsed in the same manner.
```
Comment (royfielding) [inline draft-ietf-httpbis-semantics-latest.xml:2176]:
```suggestion
   prior messages). However, a client &MUST; retain knowledge of the request when
   parsing, interpreting, or caching a corresponding response. For example,
   responses to the <x:ref>HEAD</x:ref> method look just like the beginning of a
   response to <x:ref>GET</x:ref>, but cannot be parsed in the same manner.
```

**Comment by royfielding** [comment]
Well, the simple version is a normative statement in sheep's clothing. Let's just make it normative.

**Review by reschke — APPROVED** [review]

**Review by mnot — APPROVED** [review]

**Review by martinthomson — APPROVED** [review]

**Pull request merged**